homeassistant/components/homeconnect/api.py

```diff
@@ -0,0 +1,412 @@
+
+
+class HomeConnectDevice:
+    """Generic Home Connect device."""
+
+    # for some devices, this is instead 'BSH.Common.EnumType.PowerState.Standby'
+    # see https://developer.home-connect.com/docs/settings/power_state
+    power_off_state = "BSH.Common.EnumType.PowerState.Off"
+
+    def __init__(self, appliance):
+        """Initialize the device class."""
+        self.appliance = appliance
+        self.entities = []
+
+    def initialize(self):
+        """Fetch the info needed to initialize the device."""
+        try:
+            self.appliance.get_status()
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch appliance status. Probably offline.")
+        try:
+            self.appliance.get_settings()
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch settings. Probably offline.")
+        try:
+            program_active = self.appliance.get_programs_active()
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch active programs. Probably offline.")
+            program_active = None
+        if program_active and "key" in program_active:
+            self.appliance.status["BSH.Common.Root.ActiveProgram"] = {
+                "value": program_active["key"]
+            }
+        self.appliance.listen_events(callback=self.event_callback)
+
+    def event_callback(self, appliance):
+        """Handle event."""
+        _LOGGER.debug("Update triggered on %s", appliance.name)
+        _LOGGER.debug(self.entities)
+        _LOGGER.debug(self.appliance.status)
+        for entity in self.entities:
+            entity.async_entity_update()
+
+
+class HomeConnectEntity(Entity):
+    """Generic Home Connect entity (base class)."""
+
+    def __init__(self, device, name):
+        """Initialize the entity."""
+        self.device = device
+        self._name = name
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the node (used for Entity_ID)."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Home Connect and the entity name."""
+        return f"{self.device.appliance.haId}-{self.name}"
+
+    @property
+    def device_info(self):
+        """Return info about the device."""
+        return {
+            "identifiers": {(DOMAIN, self.device.appliance.haId)},
+            "name": self.device.appliance.name,
+            "manufacturer": self.device.appliance.brand,
+            "model": self.device.appliance.vib,
+        }
+
+    @callback
+    def async_entity_update(self):
+        """Update the entity."""
+        _LOGGER.debug("Entity update triggered on %s", self)
+        self.async_schedule_update_ha_state(True)
+
+
+class DeviceWithPrograms(HomeConnectDevice):
+    """Device with programs."""
+
+    _programs = []
```

Comment:
Can we make this an instance attribute instead? Mutable objects as class attributes are sneaky. Do we really need that?

If this attribute should be considered constant we should show that by writing it in all capital letters.

Reply:
Indeed, they are constants. Capitalized.